# Evaluating My Triathlon Performance in Races

## 2023 Leidedorp OD

In [298]:
import pandas as pd

Downloading the data

In [299]:
df = pd.read_html('https://nl.mylaps.com/evenementen/uitslagen/2023/aug/20/leiderdorp/STA_Tus.html')

Pre-processing the data

In [300]:
df[0].columns = df[0].iloc[4]

In [301]:
df = df[0].iloc[5:-3]

In [303]:
def stringtime_to_seconds(t):
    if len(t.split(':')) == 2:
        return int(t.split(':')[-1]) + int(t.split(':')[-2])*60 
    if len(t.split(':')) == 3:
        return int(t.split(':')[-1]) + int(t.split(':')[-2])*60 +  int(t.split(':')[-3])*3600 
    else:
        return 0

df['Zwem_total_sec'] = df['Zwem'].astype(str).apply(stringtime_to_seconds)    
df['Wis1_total_sec'] = df['Wis1'].astype(str).apply(stringtime_to_seconds)
df['Fiets_total_sec'] = df['Fiets'].astype(str).apply(stringtime_to_seconds)
df['Wis2_total_sec'] = df['Wis2'].astype(str).apply(stringtime_to_seconds)
df['Loop_total_sec'] = df['Loop'].astype(str).apply(stringtime_to_seconds)

In [304]:
df['Zwem_speed_per100m'] = pd.to_datetime(df['Zwem_total_sec']/15,unit='s').dt.time

df['Fiets_speed_km/h'] = 40/(df['Fiets_total_sec']/3600)

df['Loop_pace_min/km'] = pd.to_datetime(df['Loop_total_sec']/10,unit='s').dt.time

Saving the data

In [ ]:
df.to_excel('races/2023_leidedorp_od.xlsx')

# Overall numbers

In [305]:
print("Total participants:",len(df))
print("Total Male:",len(df[df['MV']=='M']))
print("Total Female:",len(df[df['MV']=='V']))
print("Didn't Finish:",len(df[df['#Tot']=='DNF']))

Total participants: 119
Total Male: 90
Total Female: 29
Didn't Finish: 2


## Comparing Myself with the Winner

In [306]:
df[(df['Naam'].str.contains('Jair')) | (df['#Tot']=='1')][['#Tot','#Cat','Zwem','Zwem_speed_per100m','Wis1','Fiets','Fiets_speed_km/h','Wis2','Loop','Loop_pace_min/km','Totaal']]

4  #Tot #Cat   Zwem Zwem_speed_per100m  Wis1    Fiets  Fiets_speed_km/h  Wis2  \
5     1    1  20:58    00:01:23.866666  1:15    58:56         40.723982  1:55   
80   76   62  36:48    00:02:27.200000  3:12  1:16:23         31.420467  3:25   

4    Loop Loop_pace_min/km   Totaal  
5   35:33  00:03:33.300000  1:58:34  
80  53:30         00:05:21  2:53:16

'#Cat': Male or Female category!

## Removing participants that didn't finish

In [307]:
df = df[df['#Tot']!='DNF']

## Comparing myself to ALL other participants (whether in q1, q2/median, q3, q4)

In [308]:
df = df.sort_values(['Zwem_total_sec']).reset_index()
df.index += 1
df.index.rename('#Zwem', inplace=True)

In [309]:
df = df.sort_values(['Wis1_total_sec']).reset_index()
df.index += 1
df.index.rename('#Wis1', inplace=True)

In [310]:
df = df.sort_values(['Fiets_total_sec']).reset_index()
df.index += 1
df.index.rename('#Fiets', inplace=True)

In [311]:
df = df.sort_values(['Wis2_total_sec']).reset_index()
df.index += 1
df.index.rename('#Wis2', inplace=True)

In [312]:
df = df.sort_values(['Loop_total_sec']).reset_index()
df.index += 1
df.index.rename('#Loop', inplace=True)

In [313]:
df = df.reset_index()

In [314]:
df[(df['Naam'].str.contains('Jair')) | (df['#Tot']=='1')][['Naam','#Tot','#Zwem','#Wis1','#Fiets','#Wis2','#Loop']]

4                            Naam #Tot  #Zwem  #Wis1  #Fiets  #Wis2  #Loop
0               Hugo Jan Bosscher    1      1      1       1      2      1
69  José Jair Cardoso de Santanna   76     82     68      76     78     70

Out of 119 participants (q1:0-30, q2:30-60, q3:60-90, q4:90-119):
- **Swim**: q3
- **T1**: q3
- **Biking**: q3
- **T2**: q3
- **Running**: q3

## Comparing myself to the Male Category
Filtering only Male

In [315]:
df = df[df['Cat']=='M']

In [316]:
df = df.sort_values(['Zwem_total_sec']).reset_index()
df.index += 1
df.index.rename('#Zwem_m', inplace=True)

In [317]:
df = df.sort_values(['Wis1_total_sec']).reset_index()
df.index += 1
df.index.rename('#Wis1_m', inplace=True)

In [318]:
df = df.sort_values(['Fiets_total_sec']).reset_index()
df.index += 1
df.index.rename('#Fiets_m', inplace=True)

In [319]:
df = df.sort_values(['Wis2_total_sec']).reset_index()
df.index += 1
df.index.rename('#Wis2_m', inplace=True)

In [320]:
df = df.sort_values(['Loop_total_sec']).reset_index()
df.index += 1
df.index.rename('#Loop_m', inplace=True)

In [321]:
df = df.reset_index()

In [322]:
df[(df['Naam'].str.contains('Jair')) | (df['#Cat']=='1')][['Naam','#Cat','#Zwem_m','#Wis1_m','#Fiets_m','#Wis2_m','#Loop_m']]

4                            Naam #Cat  #Zwem_m  #Wis1_m  #Fiets_m  #Wis2_m  \
0               Hugo Jan Bosscher    1        1        1         1        2   
57  José Jair Cardoso de Santanna   62       62       50        63       61   

4   #Loop_m  
0         1  
57       58

Out of 90 male participants (q1:0-23, q2:23-46, q3:46-68, q4:68-90):
- **Swim**: q3
- **T1**: q3
- **Biking**: q3
- **T2**: q3
- **Running**: q3

# To aim for the next race(s)

What were the speeds of the slow and fast people in q1, q2, q3, and q4?

In [356]:
modality = '#Zwem_m'
speed = 'Zwem_speed_per100m'

print(str(df[(df['Naam'].str.contains('Jair'))][speed].values[0]))

df[(df[modality]==1) |(df[modality]==23)|(df[modality]==46)|(df[modality]==68)|(df[modality]==88)][[modality, speed]].sort_values([modality]) 

00:02:27.200000


4   #Zwem_m Zwem_speed_per100m
0         1    00:01:23.866666
54       23    00:02:04.333333
39       46    00:02:17.933333
45       68    00:02:33.200000
74       88    00:03:02.066666

In [357]:
modality = '#Fiets_m'
speed = 'Fiets_speed_km/h'

print(str(df[(df['Naam'].str.contains('Jair'))][speed].values[0]))

df[(df[modality]==1) |(df[modality]==23)|(df[modality]==46)|(df[modality]==68)|(df[modality]==88)][[modality, speed]].sort_values([modality]) 

31.420466943050403


4   #Fiets_m  Fiets_speed_km/h
0          1         40.723982
19        23         35.590707
33        46         32.690125
65        68         30.934479
86        88         25.347650

In [358]:
modality = '#Loop_m'
speed = 'Loop_pace_min/km'

print(str(df[(df['Naam'].str.contains('Jair'))][speed].values[0]))

df[(df[modality]==1) |(df[modality]==23)|(df[modality]==46)|(df[modality]==68)|(df[modality]==88)][[modality, speed]].sort_values([modality]) 

00:05:21


4   #Loop_m Loop_pace_min/km
0         1  00:03:33.300000
22       23  00:04:30.700000
45       46  00:05:00.900000
67       68  00:05:47.500000
87       88  00:07:41.500000